### PR TITLE
Implement grpc server

### PR DESCRIPTION
**Pull request by TanmayArya-1p**
Addresses #33

### CODE REVIEW

**Comment by 4adex** [comment]
run cargo clippy and solve the formatting issues

**Comment by Arshdeep54** [comment]
Write unit tests for checking the connections with the server (wrong auth, major edge cases, correct connection)

File: crates/grpc_server/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+    async fn search_points(
+        &self,
+        request: Request<SearchRequest>,
+    ) -> Result<Response<SearchResponse>, Status> {
+        log_rpc("search_points", self.logging);
+
+        let search_request = request.into_inner();
+
+        //TODO: distance function panics if different dimensions; no dimension enforcement
```
Comment (4adex) [inline crates/grpc_server/src/main.rs]: Check for dimension, if different return an error to prevent panic

File: crates/grpc_server/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+        }))
+    }
+
+    async fn delete_point(&self, request: Request<PointId>) -> Result<Response<()>, Status> {
+        log_rpc("delete_point", self.logging);
+
+        let point_id = request.into_inner().id;
+
+        // TODO: delete call needs to return a boolean indicating if point is present or not
+        if self.vector_db.delete(point_id).is_err() {
```
Comment (4adex) [inline crates/grpc_server/src/main.rs]: vector db returns true or false, you can implement this

File: crates/grpc_server/src/constants.rs [hunk 3]
```diff
@@ -0,0 +1,14 @@
+use defs::Similarity;
+use defs::Similarity::{Cosine, Euclidean, Hamming, Manhattan};
+pub const ENV_HOST: &str = "GRPC_SERVER_HOST";
+pub const ENV_PORT: &str = "GRPC_SERVER_PORT";
+pub const ENV_ROOT_PASSWORD: &str = "GRPC_SERVER_ROOT_PASSWORD";
+pub const ENV_STORAGE_TYPE: &str = "GRPC_SERVER_STORAGE_TYPE";
+pub const ENV_INDEX_TYPE: &str = "GRPC_SERVER_INDEX_TYPE";
+pub const ENV_DIMENSION: &str = "GRPC_SERVER_DIMENSION";
+pub const ENV_DATA_PATH: &str = "GRPC_SERVER_DATA_PATH";
+pub const ENV_LOGGING: &str = "GRPC_SERVER_LOGGING";
```
Comment (Arshdeep54) [inline crates/grpc_server/src/constants.rs:10]: Use same varible names as strings

File: crates/grpc_server/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        }))
+    }
+
+    async fn get_point(&self, request: Request<PointId>) -> Result<Response<Point>, Status> {
+        log_rpc("get_point", self.logging);
+
+        let inner_request = request.into_inner();
+
+        let point_id = inner_request.id;
+        let point_opt = self.vector_db.get(point_id).unwrap();
```
Comment (4adex) [inline crates/grpc_server/src/main.rs]: Try avoiding unwraps where it can be.
Here you could have written it like this instead
```rust
let point_opt = self
            .vector_db
            .get(point_id)
            .map_err(|e| Status::aborted(format!("Error from database {:?}", e)))?;
 ```

File: crates/grpc_server/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+        // return error if not found
+        if point_opt.is_none() {
+            return Err(Status::not_found(format!("point not found: {}", point_id)));
```
Comment (4adex) [inline crates/grpc_server/src/main.rs]: instead of `is_none()` check, do it this way

```rust
if let Some(point) = point_opt {
            Ok(Response::new(Point {
                ... }),
                payload: None,
            }))
        } else {
            return Err(Status::not_found(format!("point not found: {point_id}")));
        }
  ```
  (check the correct status here, giving this as an example)

**Review by 4adex — COMMENTED** [review]
Do these with tests and linting fix, else looks good.

**Comment by TanmayArya-1p** [comment]
fixed #45

**Comment by TanmayArya-1p** [comment]
@4adex  @Arshdeep54  Added tests,fixed clippy warnings and the unwrap issues.

**Comment by Arshdeep54** [comment]
Squash into a single commit.

**Comment by TanmayArya-1p** [comment]
@Arshdeep54 squashed commits
Sorry didn't see your message earlier

**Comment by TanmayArya-1p** [comment]
Requires #55 to be merged for CI to pass

File: crates/grpc_server/src/service.rs [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        let search_request = request.into_inner();
+
+        // extract request arguments
+        let query_vect = search_request
+            .query_vector
+            .ok_or(Status::invalid_argument("Invalid query_vector"))?;
+        let similarity = SIMILARITY_PROTOBUFF_MAP
+            .get(search_request.similarity as usize)
+            .ok_or(Status::internal("Invalid similarity"))?;
+        let limit = search_request.limit;
```
Comment (Arshdeep54) [inline crates/grpc_server/src/service.rs:111]: add limit>=1 check and return invalid arguments if not.
You can cap it to MAX_LIMIT as well.

File: crates/grpc_server/src/service.rs [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+        let search_request = request.into_inner();
+
+        // extract request arguments
+        let query_vect = search_request
+            .query_vector
+            .ok_or(Status::invalid_argument("Invalid query_vector"))?;
+        let similarity = SIMILARITY_PROTOBUFF_MAP
+            .get(search_request.similarity as usize)
+            .ok_or(Status::internal("Invalid similarity"))?;
```
Comment (Arshdeep54) [inline crates/grpc_server/src/service.rs]: Change this error type to invalid args

File: crates/grpc_server/src/service.rs [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    async fn get_point(&self, request: Request<PointId>) -> Result<Response<Point>, Status> {
+        log_rpc("get_point", self.logging);
+
+        let inner_request = request.into_inner();
+
+        let point_id = inner_request.id;
+        let point_opt = self
+            .vector_db
+            .get(point_id)
+            .map_err(|e| Status::aborted(format!("point not found {:?}", e)))?;
```
Comment (Arshdeep54) [inline crates/grpc_server/src/service.rs:78]: Change this to internal error

File: crates/grpc_server/src/interceptors.rs [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+use tracing::{Level, event};
+
+#[derive(Clone)]
+pub struct AuthInterceptor {
+    root_password: String,
+}
+
+impl Interceptor for AuthInterceptor {
+    fn call(&mut self, req: tonic::Request<()>) -> Result<tonic::Request<()>, Status> {
+        let auth_token = match req.metadata().get("authorization") {
```
Comment (Arshdeep54) [inline crates/grpc_server/src/interceptors.rs]: Add constants for authorization and Bearer strings; these are also used multiple times in tests

**Review by Arshdeep54 — CHANGES_REQUESTED** [review]

**Review by Arshdeep54 — APPROVED** [review]
LGTM

**Comment by 4adex** [comment]
@TanmayArya-1p rebase this with refactor and resolve conflicts.
